File: dart/lib/src/utils/verify_record_staleness.dart
import '../constants/records.dart';
import '../domain/get_domain_owner.dart';
import '../record/get_record_v2_address.dart';
import '../rpc/rpc_client.dart';
import '../states/record.dart';
import '../types/validation.dart';

/// Verify record staleness synchronously
///
/// This function mirrors js-kit/src/record/verifyRecordStaleness.ts _verifyStalenessSync
///
/// [domainOwner] - The domain owner address
/// [state] - The record state to verify
///
/// Returns true if the record passes staleness validation
bool verifyStalenessSync({
  required String domainOwner,
  required RecordState state,
}) {
  final stalenessId = state.getStalenessId();
  final domainOwnerBytes = _base58Decode(domainOwner);

  return _uint8ArraysEqual(domainOwnerBytes, stalenessId.toList()) &&
      state.header.stalenessValidation == Validation.solana.index;
}

/// Parameters for verifying record staleness
class VerifyRecordStalenessParams {
  const VerifyRecordStalenessParams({
    required this.rpc,
    required this.domain,
    required this.record,
  });

  /// The RPC client for blockchain interaction
  final RpcClient rpc;

  /// The domain under which the record resides
  final String domain;

  /// The record to verify
  final Record record;
}

/// Verify record staleness asynchronously
///
/// This function mirrors js-kit/src/record/verifyRecordStaleness.ts verifyRecordStaleness
///
/// [params] - Parameters containing RPC client, domain, and record type
///
/// Returns true if the record is stale, false otherwise
Future<bool> verifyRecordStaleness(VerifyRecordStalenessParams params) async {
  final recordAddress = await getRecordV2Address(GetRecordV2AddressParams(
    domain: params.domain,
    record: params.record,
  ));

  final results = await Future.wait([
    getDomainOwner(GetDomainOwnerParams(
      rpc: params.rpc,
      domain: params.domain,
    )),
    RecordState.retrieve(params.rpc, recordAddress),
  ]);

  final domainOwner = results[0] as String;
  final state = results[1] as RecordState;

  return verifyStalenessSync(
    domainOwner: domainOwner,
    state: state,
  );
}

/// Helper function to check if two byte arrays are equal
bool _uint8ArraysEqual(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

/// Base58 decode helper
List<int> _base58Decode(String input) {
  const alphabet = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

  if (input.isEmpty) return [];

  // Count leading zeros
  var leadingZeros = 0;
  for (var i = 0; i < input.length; i++) {
    if (input[i] == '1') {
      leadingZeros++;
    } else {
      break;
    }
  }

  // Decode base58
  var decoded = BigInt.zero;
  final base = BigInt.from(58);

  for (var i = leadingZeros; i < input.length; i++) {
    final char = input[i];
    final index = alphabet.indexOf(char);
    if (index == -1) {
      throw ArgumentError('Invalid base58 character: $char');
    }
    decoded = decoded * base + BigInt.from(index);
  }

  // Convert to bytes
  final bytes = <int>[];
  while (decoded > BigInt.zero) {
    bytes.insert(0, (decoded % BigInt.from(256)).toInt());
    decoded = decoded ~/ BigInt.from(256);
  }

  // Add leading zeros
  for (var i = 0; i < leadingZeros; i++) {
    bytes.insert(0, 0);
  }

  return bytes;
}

File: js-kit/src/record/verifyRecordStaleness.ts
import {
  Address,
  GetAccountInfoApi,
  GetSlotApi,
  GetTokenLargestAccountsApi,
  Rpc,
} from "@solana/kit";

import { addressCodec } from "../codecs";
import { _getSnsDomainOwner } from "../domain/getSnsDomainOwner";
import { getRecordV2Address } from "../record/getRecordV2Address";
import { RecordState } from "../states/record";
import { Record } from "../types/record";
import { Validation } from "../types/validation";
import { assertTldSupported } from "../utils/assertTldSupported";
import { uint8ArraysEqual } from "../utils/uint8Array/uint8ArraysEqual";

/**
 * Internal helper that verifies a record's staleness validation.
 *
 * @param params Staleness verification parameters
 * @param params.domainOwner Current owner of the domain
 * @param params.state Record state to verify
 * @returns True if the record's staleness validation passes, false otherwise.
 *
 * @example
 * ```ts
 * const valid = _verifyStalenessSync({ domainOwner, state });
 * ```
 */
export const _verifyStalenessSync = ({
  domainOwner,
  state,
}: {
  domainOwner: Address;
  state: RecordState;
}) => {
  const stalenessId = state.getStalenessId();

  return (
    uint8ArraysEqual(addressCodec.encode(domainOwner), stalenessId) &&
    state.header.stalenessValidation === Validation.Solana
  );
};

/**
 * Parameters for verifying record staleness.
 *
 * @example
 * ```ts
 * const params: VerifyRecordStalenessParams = {
 *   rpc,
 *   domain: "example.sns",
 *   record: Record.Url,
 * };
 * ```
 */
export interface VerifyRecordStalenessParams {
  /** RPC client. */
  rpc: Rpc<GetAccountInfoApi & GetTokenLargestAccountsApi & GetSlotApi>;
  /** Full domain name. */
  domain: string;
  /** Record type. */
  record: Record;
}

/**
 * Verifies a record's staleness validation.
 *
 * @param params Staleness verification parameters
 * @param params.rpc RPC client implementing account and token-largest-account APIs
 * @param params.domain Full domain name including a `.sns` or `.sol` suffix
 * @param params.record Record type to verify
 * @returns True if the record's staleness validation passes, false otherwise.
 *
 * @example
 * ```ts
 * const valid = await verifyRecordStaleness({ rpc, domain: "example.sns", record: Record.Url });
 * ```
 */
export const verifyRecordStaleness = async ({
  rpc,
  domain,
  record,
}: VerifyRecordStalenessParams): Promise<boolean> => {
  const [trimmedDomain] = await assertTldSupported({ rpc, domain });
  const [domainOwner, state] = await Promise.all([
    _getSnsDomainOwner({ rpc, domain: trimmedDomain }),
    getRecordV2Address({ domain: trimmedDomain, record }).then((address) =>
      RecordState.retrieve(rpc, address)
    ),
  ]);

  return _verifyStalenessSync({ domainOwner, state });
};
